### PR TITLE
Enhance Quantity Class with Array Support and String Parser

### PR DESCRIPTION
**Pull request by atxy-blip**
## Overview
This PR is intended for enhancing the `Quantity` class to support both scalar and array values while maintaining full backward compatibility. It also introduces a string parser for convenient quantity initialization.

### 1. Unified Quantity Class
- Single `Quantity` class handles both scalar and array values
- Internal storage uses numpy arrays for unified handling
- Full backward compatibility with existing code

### 2. Array Support
- Create quantities with lists: `Quantity([1, 2, 3], "cm^{-1}")`
- Array operations: indexing, slicing, iteration
- Element-wise arithmetic operations

### 3. Enhanced Numpy Compatibility
- `shape` property for array-like access
- `__array__` method for seamless numpy conversion
- Proper handling of numpy scalar types

### 4. String Parser
- Facilitate import from config files like YAML or TOML
- Parse quantities from strings: `parse_quantity_str("100 cm^{-1}")`
- Support for both scalar and array values: `parse_quantity_str("100,200,300 eV")`

### Basic Quantity Operations
```python
# Scalar quantities (backward compatible)
energy = Quantity(100, "cm^{-1}")
temp = Quantity(300, "K")

# Array quantities (new feature)
frequencies = Quantity([100, 200, 300], "cm^{-1}")

# Unit conversion
energy_ev = energy.as_unit("eV")
frequencies_au = frequencies.as_au()

# Temperature to beta
beta = temp.to_beta()
```

### String Parser
```python
# Parse scalar quantities
q1 = parse_quantity_str("100 cm^{-1}")
q2 = parse_quantity_str("300 K")

# Parse array quantities  
q3 = parse_quantity_str("100,200,300 eV")
q4 = parse_quantity_str("1.5, 2.0, 3.5 a.u.")

# Case insensitive units
q5 = parse_quantity_str("100 cm-1")  # Alternative notation
```

### Array Operations
```python
# Create array quantity
spectrum = Quantity([100, 150, 200], "cm^{-1}")

# Access elements
first_peak = spectrum[0]  # Returns Quantity(100, "cm^{-1}")

# Iteration
for peak in spectrum:
    print(peak.as_unit("eV"))

# Shape property
print(spectrum.shape)  # (3,)
```

## Backward Compatibility
- All existing code continues to work without modification
- `.value` property works for scalar quantities
- `.values` property provides array access
- Arithmetic operations maintain same behavior

## Testing
Comprehensive test suite covers:
- Basic creation and unit conversion
- Arithmetic operations with various types
- Array operations and numpy compatibility
- String parser functionality
- Edge cases and error conditions

### CODE REVIEW

**Comment by jjren** [comment]
:thumbsup:

File: renormalizer/utils/quantity.py [hunk 1]
```diff
@@ -28,76 +29,271 @@
 
 
 def convert_to_au(num, unit):
-    assert unit in allowed_units
```
Comment (jiangtong1000) [inline renormalizer/utils/quantity.py:31]: just a quick glance 👍🏻, I feel this line might still be needed.

Comment (jiangtong1000) [reply renormalizer/utils/quantity.py:31]: oh never mind, it is handled elsewhere

**Comment by liwt31** [comment]
The CI failure is a trivial numerical inaccuracy in the order of 1e-8.

Many thanks and merged.
